### ISSUE TITLE
PGoApi is not thread-safe but it being used as such (missing pokemon on map)

### ISSUE THREAD

**Issue by DeirhX** [AHAAAAAAA/PokemonGo-Map#1897]
Increase to concurrency from https://github.com/AHAAAAAAA/PokemonGo-Map/pull/1835 now reveals that PGoApi is not thread-safe but is used concurrently to generate and dispatch requests. This currently leads into following behavior:
Threads 1 to N enter send_map_requests() simultaneously
Threads 1 to N call api.set_position simultaneously and overwrite the one shared value
Threads 1 to N call api.get_map_objects simultaneously and each of them adds its request. api now contains N such requests.
Threads 1 to N call api.call simultaneously which means the N requests get dispatched N times [O(N*N) !] all the while sharing the same position.
Chaos ensues, some cells are skipped since call() uses the position set by set_position which was overwritten by the last thread.

## Expected Behavior
Concurrent requests work the same way as if were sent serially

## Current Behavior
PGoApi keeps internal state between constructing and sending requests, opening a window to any concurrently running thread to overwrite it. This leads to missing **pokemon on map** since some cell requests are skipped due to api's internal position being overwritten by another thread before the RPC call is dispatched.

## Possible Solution
PGoApi needs to be thread-safe or immutable after login. Sharing login state between all requests definitely makes sense, sharing position does not.

**Comment by DeirhX** [comment]
I suggest either separating login state and request construction into two classes (login state being immutable so it doesn't have to be locked when checking if already logged in) or removing all the other internal state from PGoApi (position, req_method_list) so multiple threads can each construct their own requests independently without affecting its shared state.

**Comment by AHAAAAAAA** [comment]
@tejado thoughts?

**Comment by DeirhX** [comment]
As easiest solution for now I would opt for splitting PGoApiRequest from PGoApi so each thread can construct its request object independently. I can code it and send a Pull Request for review.

**Comment by AHAAAAAAA** [comment]
Good idea, I'd love to test that out. It would be great if you can do a PR against the main pgoapi repo too!

**Comment by DeirhX** [comment]
Sure thing. Should be ready shortly, I hope.

**Comment by Chlodochar** [comment]
Sounds good. It probably doesn't need changes to pgoapi itself right? Otherwise, please consider merging #671.

**Comment by DeirhX** [comment]
I would need to some split functions of pgoapi into pgoapirequest (all the req_method_list and position functionality) while leaving the login functionality in pgoapi. Pgoapi would then act as an existing connection/login to the server, allowing you to create new pgoapirequest object each time you want to request something from the server.

**Comment by DeirhX** [comment]
Since login can only expire, it can be then treated as immutable object.

**Comment by Chlodochar** [comment]
Might be good to mention @tejado here.

Using `api._auth_provider.get_ticket` and `api._auth_provider.set_ticket` might do the trick. Login with one PGoApi and let other PGoApis copy the _auth_provider and/or ticket.

**Comment by DeirhX** [comment]
Yup, also possible. However the idea of 'copying an api' sounds strange to me when it's possible to skip the copy entirely and share the login state.

**Comment by Chlodochar** [comment]
Indeed. But that would (probably) require altering pgoapi. I'm somewhat against changing pgoapi inside PokemonGo-Map due to #671 that still needs to be merged ;).

**Comment by DeirhX** [comment]
I understand. Any timeline on that?  
We can make a hotfix the way you say for now and refactor pgoapi after that.

**Comment by Chlodochar** [comment]
Sounds great! #671 needs more eyes from collaborators I guess. For me the timeline is ASAP (I've been rebasing that one for 4 days. 4 years in PokemonGoDev time.

**Comment by DeirhX** [comment]
Ouch, my condolences go with you for your rebasing .)  
I'll prepare the hotfix and full refactor PRs and let's review it together. Refactor is almost ready and does seem to fix the missing pokemon as expected.

**Comment by DeirhX** [comment]
Also scan speedup time is expected since if many requests met on the same call to api.call()....N*N redundant identical calls really stalled the server.  Happened to me and it was from investigating this strange overload that has led me to discovering this ;)

**Comment by DeirhX** [comment]
Created the hotfix PR here: https://github.com/AHAAAAAAA/PokemonGo-Map/pull/1913  It's a hack and it pains me too see it ;)

When possible (after merging in https://github.com/AHAAAAAAA/PokemonGo-Map/pull/671), please replace by full refactor here: https://github.com/AHAAAAAAA/PokemonGo-Map/pull/1910  that doesn't copy things around randomly ;)

**Comment by tejado** [comment]
Hi all, thanks. Have thread-safe also on my "roadmap". I will check #1910 later today. If it looks good for me, I will integrate it for my next pgoapi 1.2.0 version.
